dlib/math/windows.h

```diff
@@ -0,0 +1,264 @@
+               0.5 * std::cos(phi) +
+               0.08 * std::cos(2.0 * phi);
+    }
+
+    // ----------------------------------------------------------------------------------------
+
+    template<typename R>
+    R blackman_nuttall(std::size_t i, std::size_t N, WindowSymmetry type)
+    /*!
+        This computes the Blackman-Nuttall window function.
+        See https://en.wikipedia.org/wiki/Window_function#Blackman%E2%80%93Nuttall_window.
+
+        requires
+            - R is float, double, or long double
+            - 0 <= i < N
+    !*/
+    {
+        static_assert(std::is_floating_point<R>::value, "template parameter must be a floating point type");
+        DLIB_ASSERT(i < N, "index out of range");
+        const std::size_t size = type == SYMMETRIC ? N-1 : N;
+        const R phi = (2.0 * pi * i) / size;
+        return 0.3635819 -
+               0.4891775 * std::cos(phi) +
+               0.1365995 * std::cos(2*phi) -
+               0.0106411 * std::cos(3*phi);
+    }
+
+    // ----------------------------------------------------------------------------------------
+
+    template<typename R>
+    R blackman_harris(std::size_t i, std::size_t N, WindowSymmetry type)
+    /*!
+        This computes the Blackman-Harris window function.
+        See https://en.wikipedia.org/wiki/Window_function#Blackman%E2%80%93Harris_window.
+
+        requires
+            - R is float, double, or long double
+            - 0 <= i < N
+    !*/
+    {
+        static_assert(std::is_floating_point<R>::value, "template parameter must be a floating point type");
+        DLIB_ASSERT(i < N, "index out of range");
+        const std::size_t size = type == SYMMETRIC ? N-1 : N;
+        const R phi = (2.0 * pi * i) / size;
+        return 0.35875 -
+               0.48829 * std::cos(phi) +
+               0.14128 * std::cos(2*phi) -
+               0.01168 * std::cos(3*phi);
+    }
+
+    // ----------------------------------------------------------------------------------------
+
+    template<typename R>
+    R blackman_harris7(std::size_t i, std::size_t N, WindowSymmetry type)
+    /*!
+        This computes the 7-order Blackman-Harris window function.
+
+        requires
+            - R is float, double, or long double
+            - 0 <= i < N
+    !*/
+    {
+        static_assert(std::is_floating_point<R>::value, "template parameter must be a floating point type");
+        DLIB_ASSERT(i < N, "index out of range");
+        const std::size_t size = type == SYMMETRIC ? N-1 : N;
+        const R phi = (2.0 * pi * i) / size;
+        return 0.27105 -
+               0.43329 * std::cos(phi) +
+               0.21812 * std::cos(2*phi) -
+               0.06592 * std::cos(3*phi) +
+               0.01081 * std::cos(4*phi) -
+               0.00077 * std::cos(5*phi) +
+               0.00001 * std::cos(6*phi);
+    }
+
+    // ----------------------------------------------------------------------------------------
+
+    enum WindowType
```

Comment:
```suggestion
    enum window_type
```
Since everything in dlib is snake_case.